port/osx/omrosdump.c

```diff
@@ -0,0 +1,388 @@
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup Port
+ * @brief User space system dump for OSX
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <mach/mach.h>
+#include <mach/mach_vm.h>
+#include <mach/machine.h>
+#include <mach/thread_status.h>
+#include <mach/vm_region.h>
+#include <mach-o/loader.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/proc.h>
+#include <sys/sysctl.h>
+#include <sys/syslimits.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "omrport.h"
+#include "omrportpriv.h"
+#include "omrosdump_helpers.h"
+
+/* current thread flavors needed for OSX debugging */
+struct thread_command_full_64 {
+	uint32_t cmd;
+	uint32_t cmdsize;
+	x86_thread_state_t thread_state;
+	x86_float_state_t float_state;
+	x86_exception_state_t exceptions;
+};
+
+static char corefile_name[PATH_MAX];
+static int corefile_fd = -1;
+
+static int coredump_to_file(mach_port_t, pid_t);
+static int list_thread_commands(mach_port_t, struct thread_command_full_64 **, natural_t *);
+static int list_segment_commands(mach_port_t, struct segment_command_64 **, natural_t *);
+
+static int coredump_to_file(mach_port_t task_port, pid_t pid) {
+	kern_return_t kr = KERN_SUCCESS;
+	struct mach_header_64 *mh64;
+	struct segment_command_64 *segments = NULL;
+	struct thread_command_full_64 *threads = NULL;
+	natural_t segment_count;
+	natural_t thread_count;
+	natural_t cpu_count;
+	processor_basic_info_t proc_info_array;
+	mach_msg_type_number_t info_count;
+	off_t file_off = 0;
+	uint64_t seg_file_off = 0;
+
+	corefile_fd = open(corefile_name, O_RDWR | O_CREAT | O_EXCL, 0600);
+
+	if (-1 == corefile_fd) {
+		perror("open()");
+		goto done;
+	}
+
+	kr = host_processor_info(mach_host_self(), PROCESSOR_BASIC_INFO, &cpu_count, (processor_info_array_t *)&proc_info_array, &info_count);
+	if (kr != KERN_SUCCESS) {
+		mach_error("failed to get processor info:\n", kr);
+		goto done;
+	}
+
+	mh64 = calloc(1, sizeof(struct mach_header_64));
+	mh64->magic = MH_MAGIC_64;
+	/* host_processor_info does not return a value with the 64-bit flag set so set manually since we only deal with 64-bit on OSX*/
+	mh64->cputype = proc_info_array[0].cpu_type | CPU_ARCH_ABI64;
+	mh64->cpusubtype = proc_info_array[0].cpu_subtype;
+	/* update ncmds and sizeofcmds after processing thread and segment commands */
+	mh64->ncmds = 0;
+	mh64->sizeofcmds = 32;
+	mh64->filetype = MH_CORE;
+
+	file_off += sizeof(struct mach_header_64);
+	segments = NULL;
+	kr = list_thread_commands(task_port, &threads, &thread_count);
+	if (kr != KERN_SUCCESS) {
+		mach_error("error getting thread command data:\n", kr);
+		goto done;
+	}
+
+	for (int i = 0; i < thread_count; i++) {
```

Comment:
`int` is the wrong type here; it should be the same as `thread_count` (i.e. `natural_t`).